Add production ready cluster setup (#981)
rancher/docs: content/rancher/v2.x/en/cluster-provisioning/production/_index.md

---
title: Production Ready Cluster
weight: 2510
---

While Rancher makes it easy to create Kubernetes clusters, a production ready cluster takes more consideration and planning. There are three roles that can be assigned to nodes: `etcd`, `controlplane` and `worker`. In the next sections each of the roles will be described in more detail.

When designing your cluster(s), you have two options:

> * Use dedicates nodes for each role. This ensures resource availability for the components needed for the specified role. Another reason can be if you want to strictly isolate network traffic between each of the roles according to the [Port Requirements]({{< baseurl >}}/rancher/v2.x/en/installation/references/).

Reviewer: ```suggestion
* Use dedicated nodes for each role. This ensures resource availability for the components needed for the specified role. It also strictly isolates network traffic between each of the roles according to the [Port Requirements]({{< baseurl >}}/rancher/v2.x/en/installation/references/).
```

> * Use the roles `etcd` and `controlplane` on the same nodes. Make sure the specifications of your nodes are according to the hardware requirements of each role.

Reviewer: ```suggestion
* Assign the `etcd` and `controlplane` roles to the same nodes. These nodes must the meet the hardware requirements for both roles.
```

>**Note:** Nodes with the `controlplane` role are shown as `Unschedulable` in the UI, meaning no pods will be scheduled to these nodes by default.

References:

* [Kubernetes: Master Components](https://kubernetes.io/docs/concepts/overview/components/#master-components)

### Hardware Requirements

Please see [Kubernetes: Building Large Clusters](https://kubernetes.io/docs/setup/cluster-large/) for the hardware requirements.

### Count of controlplane Nodes

Adding more than one node with the `controlplane` role makes every master component highly available. See below for a breakdown of how high availability is achieved per component.

#### kube-apiserver

The Kubernetes API server (`kube-apiserver`) scales horizontally. Each node with the role `controlplane` will be added to the NGINX proxy on the nodes with components that need to access the Kubernetes API server. This means that if a node becomes unreachable, the local NGINX proxy on the node will forward the request to another Kubernetes API server in the list.

#### kube-controller-manager

> The Kubernetes controller manager uses leader election using a lease in etcd. Other instances will see an active leader and wait for that lease to expire (when a node is unresponsive) to elect a new leader.

Reviewer: We need to define what a leader is.

#### kube-scheduler

The Kubernetes scheduler uses leader election using a lease in etcd. Other instances will see an active leader and wait for that lease to expire (when a node is unresponsive) to elect a new leader.

## worker

Nodes with the `worker` role run the Kubernetes node components. See [Kubernetes: Node Components](https://kubernetes.io/docs/concepts/overview/components/#node-components) for a detailed list of components.

References:

* [Kubernetes: Node Components](https://kubernetes.io/docs/concepts/overview/components/#node-components)

### Hardware Requirements

The hardware requirements for nodes with the `worker` role mostly depend on your workloads. The minimum to run the Kubernetes node components is 1 CPU (core) and 1GB of memory.

### Count of worker Nodes

Adding more than one node with the `worker` role will make sure your workloads can be rescheduled if a node fails.

## Networking

> Cluster nodes should be located within a single region. Most cloud providers provide multiple availability zones within a region, which can be used to create higher availability for your cluster. Using multiple availability zones is fine for nodes with any role. If you are using [Kubernetes Cloud Provider]({{< baseurl >}}/rancher/v2.x/en/cluster-provisioning/rke-clusters/options/cloud-providers/) resources, consult the documentation for any restrictions (i.e. zone storage restrictions)

Reviewer: ```suggestion
Cluster nodes should be located within a single region. Most cloud providers provide multiple availability zones within a region, which can be used to create higher availability for your cluster. Using multiple availability zones is fine for nodes with any role. If you are using [Kubernetes Cloud Provider]({{< baseurl >}}/rancher/v2.x/en/cluster-provisioning/rke-clusters/options/cloud-providers/) resources, consult the documentation for any restrictions (i.e. zone storage restrictions).
```

## Cluster Diagram

This diagram is applicable to Kubernetes clusters built using RKE or [Rancher Launched Kubernetes]({{< baseurl >}}/rancher/v2.x/en/cluster-provisioning/rke-clusters/).

> ![Cluster diagram]({{< baseurl >}}/img/rancher/clusterdiagram.svg)

Reviewer: We should add a sentence that describes what the lines represent and if their colors have any significance.

## Production checklist

* Nodes should have one of the following role configurations:
  * `etcd`
  * `controlplane`
  * `etcd` and `controlplane`

>   * `worker` (The role `worker` should not be used on a node with the role `etcd` or `controlplane`)

Reviewer: ```suggestion
  * `worker` (the role `worker` should not be used on a node with the role `etcd` or `controlplane`)
```

* Network traffic is only strictly allowed according to [Port Requirements]({{< baseurl >}}/rancher/v2.x/en/installation/references/).

> * Have at least three nodes with the role `etcd` to survive losing one node. Increase for higher node fault toleration, spread across (availability) zones to provide even better fault tolerance.

Reviewer: ```suggestion
* Have at least three nodes with the role `etcd` to survive losing one node. Increase this count for higher node fault toleration, and spread them across (availability) zones to provide even better fault tolerance.
```

> * Have at least two nodes with the role `controlplane` to make the Kubernetes master components highly available.

Reviewer: Assign two or more nodes the `controlplane` role for cluster high availability.

> * Have at least two nodes with the role `worker` to make it possible to reschedule your workloads on node failure.

Reviewer: Assign two or more nodes the `worker` role for workload rescheduling upon node failure.

> * Have etcd snapshots enabled, check the created snapshots and run a disaster recovery scenario.

Reviewer: Enable etcd snapshots. Verify that snapshots are being created, and run a disaster recovery scenario to verify the snapshots are valid.

> * Load tests have been performed on the cluster according to your workloads. This will ensure the hardware requirements are adequate for your workloads.

Reviewer: Perform load tests on your cluster to verify that its hardware can support your workloads.

> * Configure Alerts/Notifiers for Kubernetes components (System Service).

Reviewer: ```suggestion
* Configure alerts/notifiers for Kubernetes components (System Service).
```

Question: why do you have "system service" in parenthesis here?

Author: That's what Kubernetes components are called in the UI.

> * Configure Logging to analyze what is going on in your cluster and for possible post-mortem analysis.

Reviewer: ```suggestion
* Configure logging for cluster analysis and ad hoc post-mortems.
```

## RKE cluster running Rancher HA

> When you setup Rancher using the [High Availability (HA) Install]({{< baseurl >}}/rancher/v2.x/en/installation/ha/), the requirements are three nodes with all roles. This fulfills the requirements for a production ready cluster, with the exception of the `worker` role that is added to the nodes. Because the cluster should be dedicated to running Rancher only (and no other workloads of any kind), this is an acceptable design. Three nodes allow for one node failure for `etcd`, it will have multiple instances of the master components by having multiple `controlplane` nodes, and because Rancher will be deployed to the cluster, it needs the `worker` role to be schedulable for workloads.

Reviewer: Can we soften the language here? I understand what you're going for, but the wording makes it sounds a little bit like we're breaking our own rules.

```suggestion
You may have noticed that our [High Availability (HA) Install]({{< baseurl >}}/rancher/v2.x/en/installation/ha/) instructions do not meet our definition of a production-ready cluster, as there are no dedicated nodes for the `worker` role. However, for your Rancher installation, this three-node cluster is valid, as:

 - It allows one `etcd`  node failure.
 - It maintains multiple instances of the master components by having multiple `controlplane` nodes 
 - Because Rancher will be deployed to the cluster, each node requires the `worker` role so that it can reschedule workloads.
```
